File: apps/cli/src/openai.zig
const std = @import("std");
const emoji = @import("emoji.zig");
const utils = @import("utils.zig");

const Message = struct {
    role: []const u8,
    content: []const u8,
};

const CompletionRequest = struct {
    messages: []Message,
    model: []const u8,
};

const Choice = struct {
    message: Message,
};

const CompletionResponse = struct {
    choices: []Choice,
};

pub const Model = enum(u2) {
    Fast,
    Default,

    /// Returns the string representation of the enum value
    pub fn toString(self: Model) []const u8 {
        return switch (self) {
            .Fast => "gpt-3.5-turbo",
            .Default => "gpt-4-turbo-preview",
        };
    }

    /// Checks whether the provided model is supported
    pub fn isSupported(model: []const u8) bool {
        return std.mem.eql(u8, model, Model.Fast.toString()) or std.mem.eql(u8, model, Model.Default.toString());
    }
};

const SYSTEM_MESSAGE =
    \\ You are a helpful coding assistant responsible for generating fitting commit messages.
    \\ You will be provided a git diff or code snippet and you are expected to provide a suitable commit message.
    \\ If a user provides anything else than you are expecting respond with a fitting message and ask for the correct input (don't include emojis in this message).
    \\
    \\ When reviewing the diff or code, focus on identifying the main purpose of the changes.
    \\ Describe the change to the best of your capabilities in a maximum of one short sentence on one line.
    \\
    \\ When reviewing a diff, pay attention to the changed filenames and extract the context of the changes.
    \\ This will help you create a more relevant and informative commit message.
    \\ Here are some examples of how you can interpret some changed filenames:
    \\  - Files or filepaths that reference testing are usually related to tests.
    \\  - Markdown files are usually related to documentation.
    \\  - Config file adjustments are usually related to configuration changes.
    \\
    \\ Try to match the generated message to a fitting emoji using its description from the provided list above.
    \\ So go look in the descriptions and find the one that best matches the description.
    \\
    \\ Always start your commit message with a gitmoji followed by the message starting with a capital letter.
    \\ Never mention filenames or function names in the message.
    \\
    \\ Below you can find a list of available gitmojis and their descriptions. Try to look for a fitting emoji and message.
    \\ Use the code representation of the emoji in the commit message.
    \\
    \\ A gitmoji commit message should look like the following: :code: Your message here
;

pub fn getCompletion(allocator: std.mem.Allocator, prompt: []const u8, model: []const u8) !CompletionResponse {
    if (!Model.isSupported(model)) {
        std.debug.print("Unsupported model provided. If you believe the model should be supported, report an issue at https://github.com/segersniels/genmoji/issues.\n", .{});
        std.process.exit(1);
    }

    var env = try std.process.getEnvMap(allocator);
    defer env.deinit();

    const api_key = env.get("OPENAI_API_KEY") orelse {
        std.debug.print("Missing OPENAI_API_KEY environment variable\n", .{});
        std.process.exit(1);
    };

    const bearer_token = try std.fmt.allocPrint(allocator, "Bearer {s}", .{api_key});
    const headers = std.http.Client.Request.Headers{ .authorization = std.http.Client.Request.Headers.Value{ .override = bearer_token }, .content_type = std.http.Client.Request.Headers.Value{ .override = "application/json" } };

    const gitmojis = try emoji.fetchGitmojis(allocator);
    const system_message = try std.fmt.allocPrint(allocator, "{s}\n\nHere's an overview of available gitmojis and when they should be used:\n{s}", .{ SYSTEM_MESSAGE, std.json.fmt(gitmojis.gitmojis, .{}) });

    var messages = std.ArrayList(Message).init(allocator);
    try messages.append(.{ .role = "system", .content = system_message });
    try messages.append(.{ .role = "user", .content = prompt });

    const request = CompletionRequest{
        .messages = messages.items,
        .model = model,
    };
    defer messages.deinit();

    const body = try std.json.stringifyAlloc(allocator, request, .{});
    const response = try utils.fetch(CompletionResponse, "https://api.openai.com/v1/chat/completions", .{
        .allocator = allocator,
        .method = .POST,
        .body = body,
        .headers = headers,
    });

    return response;
}

File: apps/cli/openai.go
package main

import (
	"context"
	"encoding/json"
	"strings"
	"sync"

	openai "github.com/sashabaranov/go-openai"
)

const SYSTEM_MESSAGE string = `
You are a helpful coding assistant responsible for generating fitting commit messages.
You will be provided a git diff or code snippet and you are expected to provide a suitable commit message.
If a user provides anything else than you are expecting respond with a fitting message and ask for the correct input (don't include emojis in this message).

When reviewing the diff or code, focus on identifying the main purpose of the changes.
Describe the change to the best of your capabilities in a maximum of one short sentence on one line.

When reviewing a diff, pay attention to the changed filenames and extract the context of the changes.
This will help you create a more relevant and informative commit message.
Here are some examples of how you can interpret some changed filenames:
 - Files or filepaths that reference testing are usually related to tests.
 - Markdown files are usually related to documentation.
 - Config file adjustments are usually related to configuration changes.

Try to match the generated message to a fitting emoji using its description from the provided list above.
So go look in the descriptions and find the one that best matches the description.

Always start your commit message with a gitmoji followed by the message starting with a capital letter.
Never mention filenames or function names in the message.

A gitmoji commit message should look like the following: :code: Your message here

Below you can find a list of available gitmojis and their descriptions. Try to look for a fitting emoji and message.
Use the code representation of the emoji in the commit message.
`

var FILES_TO_IGNORE = []string{
	"package-lock.json",
	"yarn.lock",
	"npm-debug.log",
	"yarn-debug.log",
	"yarn-error.log",
	".pnpm-debug.log",
	"Cargo.lock",
	"Gemfile.lock",
	"mix.lock",
	"Pipfile.lock",
	"composer.lock",
	"go.sum",
}

func splitDiffIntoChunks(diff string) []string {
	split := strings.Split(diff, "diff --git")[1:]
	for i, chunk := range split {
		split[i] = strings.TrimSpace(chunk)
	}

	return split
}

func removeLockFiles(chunks []string) []string {
	var wg sync.WaitGroup

	filtered := make(chan string)

	for _, chunk := range chunks {
		wg.Add(1)

		go func(chunk string) {
			defer wg.Done()
			shouldIgnore := false
			header := strings.Split(chunk, "\n")[0]

			// Check if the first line contains any of the files to ignore
			for _, file := range FILES_TO_IGNORE {
				if strings.Contains(header, file) {
					shouldIgnore = true
				}
			}

			if !shouldIgnore {
				filtered <- chunk
			}
		}(chunk)
	}

	go func() {
		wg.Wait()
		close(filtered)
	}()

	var result []string
	for chunk := range filtered {
		result = append(result, chunk)
	}

	return result
}

// Split the diff in chunks and remove any lock files to save on tokens
func prepareDiff(diff string) string {
	chunks := splitDiffIntoChunks(diff)

	return strings.Join(removeLockFiles(chunks), "\n")
}

type OpenAI struct {
	ApiKey string
}

func NewOpenAI(apiKey string) *OpenAI {
	return &OpenAI{
		ApiKey: apiKey,
	}
}

func (o *OpenAI) GetChatCompletion(diff string) (string, error) {
	gitmojis, err := fetchGitmojis()
	if err != nil {
		return "", err
	}

	list, err := json.Marshal(gitmojis)
	if err != nil {
		return "", err
	}

	client := openai.NewClient(o.ApiKey)
	resp, err := client.CreateChatCompletion(
		context.Background(),
		openai.ChatCompletionRequest{
			Model: CONFIG.Data.Model,
			Messages: []openai.ChatCompletionMessage{
				{
					Role:    openai.ChatMessageRoleSystem,
					Content: SYSTEM_MESSAGE + string(list),
				},
				{
					Role:    openai.ChatMessageRoleUser,
					Content: prepareDiff(diff),
				},
			},
		},
	)

	if err != nil {
		return "", err
	}

	return resp.Choices[0].Message.Content, nil
}
